fuse/file.go

```diff
@@ -3,103 +3,295 @@ package fuse
+	size uint64
+}
 
 func newFile(p *dir, e plugin.Entry) *file {
-	return &file{newFuseNode("f", p, e)}
+	return &file{fuseNode: newFuseNode("f", p, e), writers: make(map[fuse.HandleID]struct{})}
+}
+
+var _ = fs.Node(&file{})
+var _ = fs.Handle(&file{})
+
+func (f *file) Attr(ctx context.Context, a *fuse.Attr) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+
+	if len(f.writers) == 0 {
+		// Fetch updated attributes.
+		entry, err := f.refind(ctx)
+		if err != nil {
+			activity.Warnf(ctx, "FUSE: Attr errored %v, %v", f, err)
+			return err
+		}
+		f.entry = entry
+	}
+
+	f.fillAttr(a)
+	activity.Record(ctx, "FUSE: Attr %v: %+v", f, *a)
+	return nil
+}
+
+func (f *file) fillAttr(a *fuse.Attr) {
+	attr := plugin.Attributes(f.entry)
+	applyAttr(a, attr, getFileMode(f.entry))
+
+	if len(f.writers) != 0 || !attr.HasSize() {
+		// Use whatever size we know locally. Retrieving content can be expensive so we settle for
+		// including size only when it's been retreived previously by opening the file.
+		a.Size = f.size
+	}
+}
+
+func getFileMode(entry plugin.Entry) os.FileMode {
+	var mode os.FileMode
+	if plugin.WriteAction().IsSupportedOn(entry) {
+		mode |= 0220
+	}
+	if plugin.ReadAction().IsSupportedOn(entry) ||
+		plugin.StreamAction().IsSupportedOn(entry) {
+		mode |= 0440
+	}
+	return mode
 }
 
+var _ = fs.NodeOpener(&file{})
+
 // Open a file for reading or writing.
 func (f *file) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenResponse) (fs.Handle, error) {
-	activity.Record(ctx, "FUSE: Open %v", f)
+	f.mux.Lock()
+	defer f.mux.Unlock()
+	activity.Record(ctx, "FUSE: Open %v: %+v", f, *req)
 
 	// Check for an updated entry in case it has static state.
-	updatedEntry, err := f.refind(ctx)
+	entry, err := f.refind(ctx)
 	if err != nil {
 		activity.Warnf(ctx, "FUSE: Open errored %v, %v", f, err)
 		return nil, err
 	}
+	f.entry = entry
 
-	var fh fileHandle
-	fh.id = f.String()
+	readable := plugin.ReadAction().IsSupportedOn(f.entry)
+	writable := plugin.WriteAction().IsSupportedOn(f.entry)
+	switch {
+	case req.Flags.IsReadOnly() && !readable:
+		activity.Record(ctx, "FUSE: Open read-only unsupported on %v", f)
+		return nil, fuse.ENOTSUP
+	case req.Flags.IsWriteOnly() && !writable:
+		activity.Record(ctx, "FUSE: Open write-only unsupported on %v", f)
+		return nil, fuse.ENOTSUP
+	case req.Flags.IsReadWrite() && (!readable || !writable):
+		activity.Record(ctx, "FUSE: Open read-write unsupported on %v", f)
+		return nil, fuse.ENOTSUP
+	}
 
-	// Initiate content request and return a channel providing the results.
-	if plugin.ReadAction().IsSupportedOn(updatedEntry) {
-		activity.Record(ctx, "FUSE: Opened %v", f)
-		if attr := plugin.Attributes(updatedEntry); !attr.HasSize() {
-			// The entry's content size is unknown so open the file in direct
-			// IO mode. This enables FUSE to still read the entry's content so
-			// that built-in tools like cat and grep still work.
+	if req.Flags.IsReadOnly() || req.Flags.IsReadWrite() {
```

Comment:
I'm not sure these are the right conditions. We definitely don't want to call `plugin.Size` on an entry that doesn't support Read. We definitely do want the size if opening it ReadOnly, to address how Vim opens it.

If you try to do a write in the middle of a file that doesn't have a Size attribute defined, then you may end up truncating the file (because we won't realize some data is missing that we need to read before writing, see the implementation of `Flush`). So if the file implements ReadAction but is opened WriteOnly, maybe we still want to determine the size. Or maybe we want to error if you try to open a file for writing and haven't explicitly set the Size attribute.